cpp/src/interop/to_arrow_device.cu

```diff
@@ -0,0 +1,681 @@
+    out->dictionary);
+}
+
+template <typename T>
+void device_buffer_finalize(ArrowBufferAllocator* allocator, uint8_t*, int64_t)
+{
+  auto* unique_buffer = reinterpret_cast<std::unique_ptr<T>*>(allocator->private_data);
+  delete unique_buffer;
+}
+
+template <typename>
+struct is_device_scalar : public std::false_type {};
+
+template <typename T>
+struct is_device_scalar<rmm::device_scalar<T>> : public std::true_type {};
+
+template <typename T>
+int set_buffer(std::unique_ptr<T> device_buf, int64_t i, ArrowArray* out)
+{
+  ArrowBuffer* buf = ArrowArrayBuffer(out, i);
+  buf->data        = reinterpret_cast<uint8_t*>(device_buf->data());
+  buf->size_bytes  = [&] {
+    if constexpr (is_device_scalar<T>::value) {
+      return sizeof(typename T::value_type);
+    } else {
+      return device_buf->size();
+    }
+  }();
+  // we make a new unique_ptr and move to it in case there was a custom deleter
+  return ArrowBufferSetAllocator(
+    buf,
+    ArrowBufferDeallocator(&device_buffer_finalize<T>,
+                           new std::unique_ptr<T>(std::move(device_buf))));
+}
+
+int initialize_array(ArrowArray* arr, ArrowType storage_type, const cudf::column& column)
+{
+  NANOARROW_RETURN_NOT_OK(ArrowArrayInitFromType(arr, storage_type));
+  arr->length     = column.size();
+  arr->null_count = column.null_count();
+  return NANOARROW_OK;
+}
+
+struct dispatch_to_arrow_device {
+  template <typename T, CUDF_ENABLE_IF(not is_rep_layout_compatible<T>())>
+  int operator()(cudf::column&&,
+                 rmm::cuda_stream_view,
+                 rmm::mr::device_memory_resource*,
+                 ArrowArray*)
+  {
+    CUDF_FAIL("Unsupported type for to_arrow_device");
+  }
+
+  template <typename T, CUDF_ENABLE_IF(is_rep_layout_compatible<T>())>
+  int operator()(cudf::column&& column,
+                 rmm::cuda_stream_view stream,
+                 rmm::mr::device_memory_resource* mr,
+                 ArrowArray* out)
+  {
+    nanoarrow::UniqueArray tmp;
+
+    const ArrowType storage_type = [&] {
+      switch (column.type().id()) {
+        case cudf::type_id::TIMESTAMP_SECONDS:
+        case cudf::type_id::TIMESTAMP_MILLISECONDS:
+        case cudf::type_id::TIMESTAMP_MICROSECONDS:
+        case cudf::type_id::TIMESTAMP_NANOSECONDS: return NANOARROW_TYPE_TIMESTAMP;
+        case cudf::type_id::DURATION_SECONDS:
+        case cudf::type_id::DURATION_MILLISECONDS:
+        case cudf::type_id::DURATION_MICROSECONDS:
+        case cudf::type_id::DURATION_NANOSECONDS: return NANOARROW_TYPE_DURATION;
+        default: return id_to_arrow_type(column.type().id());
+      }
+    }();
+
+    NANOARROW_RETURN_NOT_OK(initialize_array(tmp.get(), storage_type, column));
+
+    auto contents = column.release();
+    if (contents.null_mask) {
+      NANOARROW_RETURN_NOT_OK(
+        set_buffer(std::move(contents.null_mask), kValidityBufferIdx, tmp.get()));
+    }
```

Comment:
Just raising this again as an options in case you missed the suggestion, but totally fine if you intentionally decided not to do it: this could also go into `initialize_array` since the bitmask handling is the same for all types AFAICT (whereas the data and children handling is not).

Reply:
I intentionally chose not to do it because of the complexities of dealing with the call to `release` versus needing the size / null_count / etc. from the column *before* calling release. I mostly didn't want to juggle the handling and lifetimes